File: week10/younggeun0/state/staticway/HasQuarterState.java
package week10.younggeun0.state.staticway;

import java.util.Random;

public class HasQuarterState extends State {
	static Random randomWinner = new Random(System.currentTimeMillis());
	@Override
	public void insertQuarter(GumballMachine gm) {
		System.out.println("동전은 한 개만 넣어주세요.");
	}
	@Override
	public void ejectQuarter(GumballMachine gm) {
		System.out.println("동전이 반환됩니다.");
		gm.setState(StaticState.noQuarterState);
	}
	@Override
	public boolean turnCrank(GumballMachine gm) {
		System.out.println("손잡이를 돌리셨습니다.");
		int winner = randomWinner.nextInt(10);
		if ((winner == 0) && (gm.getCount() > 1)) {
			gm.setState(StaticState.winnerState);
			return true;
		} else {
			gm.setState(StaticState.soldState);
			return true;
		}
	}
	@Override
	public void dispense(GumballMachine gm) {
		System.out.println("알맹이가 나갈 수 없습니다.");
	}
}

File: week10/ch.Choe/Week10/Week10/HasQuarterState.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Week10
{
    class HasQuarterState : State
    {
        Random ran = new Random();

        GumballMachine m_GumballMachine;

        public HasQuarterState(GumballMachine machine)
        {
            m_GumballMachine = machine;
        }
        public void Dispense()
        {
            Console.WriteLine("동전을 넣으셨습니다.");
        }

        public void EjectQuarter()
        {
            Console.WriteLine("동전이 반환됩니다.");
            m_GumballMachine.SetState(GumballMachine.StateType.NoQuarter);
        }

        public void InsertQuarter()
        {
            Console.WriteLine("동전을 한 개만 넣어주세요.");
        }

        public void TurnCrack()
        {
            Console.WriteLine("손잡이를 돌리셨습니다.");
            int num = ran.Next(0, 10);

            if(num == 0 && m_GumballMachine.Getcount() > 1)
            {
                m_GumballMachine.SetState(GumballMachine.StateType.Winner);
            }
            else
            {
                m_GumballMachine.SetState(GumballMachine.StateType.Sold);
            }
        }
    }
}
